### DISCUSSION TITLE
ESP.deepsleep(0) possible on ESP-01 / ESP8266? Maybe as "Restart 4" option?

### DISCUSSION THREAD

**Discussion by tjareson** [arendst/Tasmota#19016] [Ideas]
Hi,
my use case are implementations like battery dependent doorbell or door sensors for garage etc. which I would like to use Tasmota for with ideally the simplest ESP-01 modules. (these 8 pin, really small boards)

Now somehow I'm searching for an option to issue a ESP.deepsleep(0) which should set a deepsleep mode, which can be only left by hardware trigger, if I understood that right. But due to this 10 second if-statement in xdrv_29_deepsleep.ino this seems to be not possible.
Using the current deepsleep setup with for instance the max. deepsleep time of 86400 seconds doesn't work out due to 2 observations:
- When waking up premature from deepsleep tasmota seems to be so smart to discover that and goes back into sleepmode immediately without even starting to connect to wifi.
- When the deepsleep wakeup time is reached, the ESP-01 module gets into a mode of consuming something around 30mA, as it obviously tries to wake-up itself, which of course doesn't work due to XPD_DCDC not connected (and it wouldn't also be necessary in my use case and it's quite a hassle to solder that on a ESP-01) It seems to stay stuck then in this stage, as it also doesn't recognize that wake-up isn't successful.

So would it be possible to allow somehow for a ESP.deepsleep(0), with the intention to only wake up on a hardware trigger and otherwise never at all?
Another solution could be to introduced a "restart 4" option which then doesn't halt the ESP (which still let it draw around 70mA) but puts it into infinite deepsleep, which would do the trick as well. I'm just not sure if then in case of wake-up from deepsleep some other time slip corrections would take over again?

**Comment by tjareson** [comment]
I've checked the source code a bit and found in support_command.ino the CmndRestart.
There I've added another variant for a restart.

```
case 4:
    ResponseCmndChar(PSTR("Activate infinite deep sleep"));
    ESP.deepSleep(0);  
    break;
```
    
Technically it does now exactly what I need. The only small issue is that I do not see the msg anymore in the web gui.
Is the reason that it doesn't get to send it anymore, as ESP.deepSleep immediately happens or maybe I didn't understand how ResponseCmndChar works? Anyone an idea how to fix that?

**Reply by barbudor** [reply]
The ESP goes to sleep immediately , it is normal that you have no more logs

I think this is an interresting approach that could be worth a PR.

**Reply by tjareson** [reply]
I will try to create one, not sure I'll get it right, but let's see.

**Reply by barbudor** [reply]
Let me know, it you don't succeed, I'll do one on your behalf

**Comment by arendst** [comment]
Just implemented in dev branch (https://github.com/arendst/Tasmota/commit/522e6cf234c84da5650882b0fcb89cc8d1e2ffba). Let's continue this thread for further comments.

**Reply by tjareson** [reply]
Thanks. So far it didn't "try" to wake up after over 3 hours. (means power consumption 0 mA still, which would be 30mA otherwise)
I guess your implementation will behave the same. Did only test the previous simple version so far.

**Reply by sfromis** [reply]
I'd suggest to compare with a test with a shorter interval, less than one hour.

**Comment by TheChatty** [comment]
Could you please elaborate on your solution (wiring, use case, ...)? Sounds like fun to play with.

**Reply by arendst** [reply]
I imagine a (reed)switch from reset input to ground which is normally open will trigger a reset once closed momentary (by a letterbox lid). Once the reset is high again the ESP will wake-up and can perform rules/script/berry commands to send a message and go to deepsleep again.

**Reply by tjareson** [reply]
The setup is exaclty like this. The simplest way would be to connect a reed switch with the reset input. But that would require that the switch will get back into open position otherwise the ESP01 will not wake up. (not sure right now if reset was triggered by neg. or pos. edge)

The more "sophisticated" approach would be to reset it via a reed switch in series with an RC combination. Something like 3.3uF + 1MOhm and make the reed switch a two way switch. At the same time then connect the reed switch with a MOSFET (like BS170 for instance) allows for connecting a GPIO with Ground depending on the position of the reed switch. So the ESP can send the position of the switch (or door) like open/close. 
There are a couple of circuits like that if you google for it.

**Reply by barbudor** [reply]
I've use such a solution on a mailbox door notifier: micro-switch (better than reed) through a RC, creates a reset pulse to awake the ESP.
I was not using Tasmota at that time but pure Arduino because of the lack of this command.
I may revisit the project now.
